lib/bpl/transformation/shadowing.rb

```diff
@@ -26,6 +26,8 @@ def shadow_copy(node)
         shadow
       end
 
+      def shadow_expr_eq(expr) "#{expr} == #{shadow_copy(expr)}" end
```

Comment:
Not sure what you mean about “strong typing”: the Node-type argument of your new definition subsumes the Identifier-type argument of the previous definition.

The `shadow_copy` function should do a full sub-expression traversal.

If you unify your `shadow_expr_eq` into `shadow_eq` I will merge your pull request :-)